src/main/java/com/google/devtools/build/lib/buildtool/ExecutionTool.java

```diff
@@ -519,11 +519,15 @@ private ModifiedFileSet startBuildAndDetermineModifiedOutputFiles(
         startLocalOutputBuild();
       }
     }
-    if (!request.getPackageOptions().checkOutputFiles
-        // Do not skip invalidation in case the output tree is empty -- this can happen
-        // after it's cleaned or corrupted.
-        && !modifiedOutputFiles.treatEverythingAsDeleted()) {
-      modifiedOutputFiles = ModifiedFileSet.NOTHING_MODIFIED;
+    if (!request.getPackageOptions().checkOutputFiles) {
+      // Do not skip output invalidation in the following cases:
+      // 1. If the output tree is empty: this can happen after it's cleaned or corrupted.
+      // 2. For a run command: so that outputs are downloaded even if they were previously built
+      //    with --remote_download_minimal. See https://github.com/bazelbuild/bazel/issues/20843.
+      if (!modifiedOutputFiles.treatEverythingAsDeleted() && !request.getCommandName()
+          .equals("run")) {
```

Comment:
I don't think I can make the check narrower, because the remote options can change between builds without invalidating the Skyframe state. (But maybe you can see a way.)